File: core/SerializeAndDeserializeBst/SerializeAndDeserializeBst.java
package com.dbc;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class SerializeAndDeserializeBst {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    private void postOrder(TreeNode node, List<Integer> res) {
        if (node != null) {
            postOrder(node.left, res);
            postOrder(node.right, res);
            res.add(node.val);
        }
    }

    public String serialize(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        postOrder(root, res);
        String resStr = res.toString();
        return resStr.substring(1, resStr.length() - 1);
    }

    private TreeNode construct(Stack<Integer> nums, int lower, int upper) {
        if (nums.isEmpty() || nums.peek() < lower || nums.peek() > upper) {
            return null;
        }
        int root_val = nums.pop();
        TreeNode root = new TreeNode(root_val);
        root.right = construct(nums, root_val, upper);
        root.left = construct(nums, lower, root_val);
        return root;
    }


    public TreeNode deserialize(String data) {
        Stack<Integer> postOrder = new Stack<>();
        for (String num : data.split(", ")) {
            if (num.equals("")) continue;
            postOrder.add(Integer.parseInt(num));
        }
        return construct(postOrder, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
}

File: core/SerializeAndDeserializeBst/SerializeAndDeserializeBst.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    def serialize(self, root: TreeNode) -> str:
        """Encodes a tree to a single string.
        """
        res = []

        def dfs(node):
            if not node:
                return
            dfs(node.left)
            dfs(node.right)
            res.append(str(node.val))

        dfs(root)

        return " ".join(res)

    def deserialize(self, data: str) -> TreeNode:
        """Decodes your encoded data to tree.
        """
        back_tree = list(map(int, data.split()))

        def dfs(lower, upper):
            if len(back_tree) == 0 or back_tree[-1] < lower or back_tree[-1] > upper:
                return None
            root_val = back_tree.pop()
            root = TreeNode(root_val)
            root.right = dfs(root_val, upper)
            root.left = dfs(lower, root_val)
            return root

        return dfs(float("-inf"), float("inf"))




